bridges/snowbridge/pallets/outbound-queue-v2/src/lib.rs

```diff
@@ -0,0 +1,408 @@
+			if !order.fee.is_zero() {
+				let ether = ether_asset(T::EthereumNetwork::get(), order.fee);
+				T::RewardPayment::pay_reward(receipt.reward_address, ether)
+					.map_err(|_| Error::<T>::RewardPaymentFailed)?;
+			}
+
+			<PendingOrders<T>>::remove(nonce);
+
+			Self::deposit_event(Event::MessageDeliveryProofReceived { nonce });
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Generate a messages commitment and insert it into the header digest
+		pub(crate) fn commit() {
+			let count = MessageLeaves::<T>::decode_len().unwrap_or_default() as u64;
+			if count == 0 {
+				return
+			}
+
+			// Create merkle root of messages
+			let root = merkle_root::<<T as Config>::Hashing, _>(MessageLeaves::<T>::stream_iter());
+
+			let digest_item: DigestItem = CustomDigestItem::SnowbridgeV2(root).into();
+
+			// Insert merkle root into the header digest
+			<frame_system::Pallet<T>>::deposit_log(digest_item);
+
+			Self::deposit_event(Event::MessagesCommitted { root, count });
+		}
+
+		/// Process a message delivered by the MessageQueue pallet
+		pub(crate) fn do_process_message(
+			_: ProcessMessageOriginOf<T>,
+			mut message: &[u8],
+		) -> Result<bool, ProcessMessageError> {
+			use ProcessMessageError::*;
+
+			// Yield if the maximum number of messages has been processed this block.
+			// This ensures that the weight of `on_finalize` has a known maximum bound.
+			ensure!(
+				MessageLeaves::<T>::decode_len().unwrap_or(0) <
+					T::MaxMessagesPerBlock::get() as usize,
+				Yield
+			);
+
+			let nonce = Nonce::<T>::get();
+
+			// Decode bytes into Message and
+			// a. Convert to OutboundMessage and save into Messages
+			// b. Convert to committed hash and save into MessageLeaves
+			// c. Save nonce&fee into PendingOrders
+			let message: Message = Message::decode(&mut message).map_err(|_| Corrupt)?;
+			let commands: Vec<OutboundCommandWrapper> = message
+				.commands
+				.clone()
+				.into_iter()
+				.map(|command| OutboundCommandWrapper {
+					kind: command.index(),
+					gas: T::GasMeter::maximum_dispatch_gas_used_at_most(&command),
+					payload: command.abi_encode(),
+				})
+				.collect();
+
+			let abi_commands: Vec<CommandWrapper> = commands
+				.clone()
+				.into_iter()
+				.map(|command| CommandWrapper {
+					kind: command.kind,
+					gas: command.gas,
+					payload: Bytes::from(command.payload),
+				})
+				.collect();
+			let committed_message = OutboundMessageWrapper {
+				origin: FixedBytes::from(message.origin.as_fixed_bytes()),
+				nonce,
+				commands: abi_commands,
+			};
+			let message_abi_encoded_hash =
+				<T as Config>::Hashing::hash(&committed_message.abi_encode());
+			MessageLeaves::<T>::append(message_abi_encoded_hash);
+
+			let outbound_message = OutboundMessage {
+				origin: message.origin,
+				nonce,
+				commands: commands.try_into().map_err(|_| Corrupt)?,
+			};
+			Messages::<T>::append(Box::new(outbound_message));
+
+			let order = PendingOrder {
```

Comment:
@yrong comments please on why we are creating a new order.

Reply:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/e02fe302e76cd82a95fcd395295fab53987a41f9